connect-inject/health_check_controller.go

```diff
@@ -0,0 +1,260 @@
+				key, err := cache.MetaNamespaceKeyFunc(newObj)
+				if err != nil {
+					// We return here, due to startup timing on probes there is a case where we receive
+					// the failed readiness probe before processing the transition from Pending to Running.
+					// When we process ObjectCreate from processNextItem() we will append an ObjectUpdate()
+					// which has the effect of setting the health status of the newly created TTL to the current state
+					c.Log.Error(fmt.Sprintf("Failed to get key from cache for pod %s", newPod.Name))
+					return
+				}
+				c.Log.Info("Add Pod: %s", key)
+				c.Queue.Add("ADD/" + key)
+				return
+			}
+			// We will only process events for PodRunning Pods
+			if newPod.Status.Phase == corev1.PodRunning {
+				// Only queue events which satisfy the condition of a pod Status Condition transition
+				// from Ready/NotReady or NotReady/Ready
+				// In this context "Ready" is the name of the Condition field and not the actual Status
+				oldPodStatus := c.getReadyStatus(oldPod)
+				newPodStatus := c.getReadyStatus(newPod)
+				// If the Pod Status has changed, we queue the newObj and set the TTL to the newObj status
+				if oldPodStatus != newPodStatus {
+					key, err := cache.MetaNamespaceKeyFunc(newObj)
+					if err != nil {
+						c.Log.Error(fmt.Sprintf("Failed to get key from cache for pod %s", newPod.Name))
+						return
+					}
+					c.Log.Info("Update pod: %s", key)
+					c.Queue.Add("UPDATE/" + key)
+					return
+				}
+			}
+		},
+	})
+}
+
+func (c *HealthCheckController) getReadyStatus(pod *corev1.Pod) corev1.ConditionStatus {
+	for _, cond := range pod.Status.Conditions {
+		if cond.Type == "Ready" {
+			return cond.Status
+		}
+	}
+	return corev1.ConditionTrue
+}
+
+// Init is used at startup to force a Reconcile phase
+func (c *HealthCheckController) Init(_ <-chan struct{}) {
+	if err := c.Handle.Init(); err != nil {
+		c.Log.Error("Error during Reconcile phase: %v", err)
+	}
+}
+
+// Run is the main path of execution for the controller loop
+func (c *HealthCheckController) Run(stopCh <-chan struct{}) {
+	c.Log.Debug("Controller.Run: initializing")
+	// Setup the Informer
+	c.setupInformer()
+	// Next setup the work queue
+	c.setupWorkQueue()
+	// Next add eventHandlers, these are responsible for defining Create/Update/Delete functionality
+	c.addEventHandlers()
+
+	// handle a panic with logging and exiting
+	defer utilruntime.HandleCrash()
+	// block new items in the Queue in case of shutdown, drain the queue and exit
+	defer c.Queue.ShutDown()
+
+	// run the Informer to start listing and watching resources
+	go c.Informer.Run(stopCh)
+
+	// do the initial synchronization (one time) to populate resources
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("error syncing cache"))
+		return
+	}
+	// run the runWorker method every second with a stop channel
+	wait.Until(c.runWorker, time.Second, stopCh)
```

Comment:
Why are we only running every 1s? If *n* pods go unhealthy will it take up to *n* seconds for their statuses to be updated?